File: backend/src/scrapers/walmart.js
import puppeteer from "puppeteer-extra";
import StealthPlugin from "puppeteer-extra-plugin-stealth";
import { JSDOM } from "jsdom";
import nwsapi from "nwsapi";

puppeteer.use(StealthPlugin());

const WALMART_SEARCH_URLS = [
  "https://www.walmart.com/search?q=best+sellers&sort=best_seller",
  "https://www.walmart.com/search?q=electronics+best+sellers&sort=best_seller",
  "https://www.walmart.com/search?q=home+best+sellers&sort=best_seller",
  "https://www.walmart.com/search?q=kitchen+best+sellers&sort=best_seller",
];

function extractProductsFromHtml(html) {
  const dom = new JSDOM(html);
  const { window } = dom;

  const nw = nwsapi(window);
  nw.configure({ IDS_DUPES: false, LIVECACHE: true, LOGERRORS: false });

  const scriptEl = nw.first("script#__NEXT_DATA__", window.document);
  if (!scriptEl) return [];

  let data;
  try {
    data = JSON.parse(scriptEl.textContent);
  } catch {
    return [];
  }

  let stacks;
  try {
    stacks = data.props.pageProps.initialData.searchResult.itemStacks;
  } catch {
    return [];
  }

  const products = [];

  for (const stack of stacks || []) {
    for (const item of stack.items || []) {
      if (item.__typename !== "Product") continue;

      const price = parseFloat(item.price);
      if (!price || price <= 0) continue;

      const priceInfo = item.priceInfo || {};
      let originalPrice = null;
      if (priceInfo.linePriceDisplay) {
        const parsed = parseFloat(priceInfo.linePriceDisplay.replace(/[$,]/g, ""));
        if (parsed > price) originalPrice = parsed;
      }

      const ratingData = item.rating || {};
      const avgRating = item.averageRating || ratingData.averageRating || null;
      const reviews = item.numberOfReviews || ratingData.numberOfReviews || null;

      const availability = item.availabilityStatusV2 || {};
      const availText = typeof availability === "object" ? availability.display || "Unknown" : "Unknown";

      let url = item.canonicalUrl || "";
      if (url && !url.startsWith("http")) url = `https://www.walmart.com${url}`;

      products.push({
        name: item.name || "Unknown",
        price,
        original_price: originalPrice,
        discount_pct: originalPrice ? Math.round((1 - price / originalPrice) * 100) : null,
        rating: avgRating ? parseFloat(String(avgRating)) : null,
        reviews_count: reviews ? parseInt(String(reviews)) : null,
        category: item.departmentName || "General",
        source: "walmart",
        url,
        image_url: item.image || null,
        seller: item.sellerName || null,
        availability: availText,
        country: "USA",
      });
    }
  }

  return products;
}

async function launchBrowser() {
  return puppeteer.launch({
    headless: "new",
    executablePath: "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome",
    args: ["--no-sandbox", "--disable-setuid-sandbox", "--disable-blink-features=AutomationControlled"],
  });
}

export async function scrapeWalmart() {
  const allProducts = [];
  const seenUrls = new Set();

  const browser = await launchBrowser();
  try {
    const page = await browser.newPage();
    await page.setUserAgent("Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36");

    for (const url of WALMART_SEARCH_URLS) {
      try {
        await page.goto(url, { waitUntil: "domcontentloaded", timeout: 30000 });
        await page.waitForSelector("script#__NEXT_DATA__", { timeout: 10000 }).catch(() => {});
        const html = await page.content();
        const products = extractProductsFromHtml(html);

        for (const p of products) {
          if (!seenUrls.has(p.url)) {
            seenUrls.add(p.url);
            allProducts.push(p);
          }
        }
        console.log(`[walmart] ${url.split("?")[0]}: found ${products.length} products`);
      } catch (err) {
        console.error(`[walmart] Error fetching ${url}:`, err.message);
      }
    }
  } finally {
    await browser.close();
  }

  return { source: "walmart", products: allProducts, status: allProducts.length > 0 ? "success" : "error" };
}

File: scraper/scrapers/walmart.py
import httpx
import json
import re
from datetime import datetime
from models import Product, ScrapeResult

WALMART_SEARCH_URLS = [
    "https://www.walmart.com/search?q=best+sellers&sort=best_seller",
    "https://www.walmart.com/search?q=electronics+best+sellers&sort=best_seller",
    "https://www.walmart.com/search?q=home+best+sellers&sort=best_seller",
    "https://www.walmart.com/search?q=kitchen+best+sellers&sort=best_seller",
]

HEADERS = {
    "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36",
    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
    "Accept-Language": "en-US,en;q=0.9",
    "Accept-Encoding": "gzip, deflate, br",
}


def _extract_products_from_html(html: str) -> list[Product]:
    match = re.search(
        r'<script id="__NEXT_DATA__"[^>]*>(.*?)</script>', html, re.DOTALL
    )
    if not match:
        return []

    data = json.loads(match.group(1))
    try:
        search_result = data["props"]["pageProps"]["initialData"]["searchResult"]
        stacks = search_result.get("itemStacks", [])
    except (KeyError, TypeError):
        return []

    products = []
    for stack in stacks:
        items = stack.get("items", [])
        for item in items:
            if item.get("__typename") != "Product":
                continue
            try:
                price = float(item.get("price", 0))
                if price <= 0:
                    continue
            except (ValueError, TypeError):
                continue

            price_info = item.get("priceInfo", {})
            line_price = price_info.get("linePriceDisplay", "")
            original_price = None
            if line_price and line_price != f"${price:.2f}":
                try:
                    original_price = float(line_price.replace("$", "").replace(",", ""))
                except (ValueError, TypeError):
                    original_price = None

            rating_data = item.get("rating", {})
            avg_rating = item.get("averageRating") or (
                rating_data.get("averageRating") if isinstance(rating_data, dict) else None
            )
            reviews = item.get("numberOfReviews") or (
                rating_data.get("numberOfReviews") if isinstance(rating_data, dict) else None
            )

            availability = item.get("availabilityStatusV2", {})
            avail_text = availability.get("display", "Unknown") if isinstance(availability, dict) else "Unknown"

            category = item.get("departmentName", "General")

            url = item.get("canonicalUrl", "")
            if url and not url.startswith("http"):
                url = f"https://www.walmart.com{url}"

            products.append(Product(
                name=str(item.get("name", "Unknown")),
                price=price,
                original_price=original_price,
                discount_pct=None,
                rating=float(avg_rating) if avg_rating else None,
                reviews_count=int(reviews) if reviews else None,
                category=str(category),
                source="walmart",
                url=url,
                image_url=item.get("image"),
                seller=item.get("sellerName"),
                availability=avail_text,
                country="USA",
            ))

    return products


async def scrape_walmart() -> ScrapeResult:
    all_products = []
    seen_urls = set()

    async with httpx.AsyncClient(follow_redirects=True, timeout=20) as client:
        for url in WALMART_SEARCH_URLS:
            try:
                resp = await client.get(url, headers=HEADERS)
                if resp.status_code == 200:
                    products = _extract_products_from_html(resp.text)
                    for p in products:
                        if p.url not in seen_urls:
                            seen_urls.add(p.url)
                            all_products.append(p)
            except Exception as e:
                print(f"[walmart] Error fetching {url}: {e}")

    return ScrapeResult(
        source="walmart",
        products=all_products,
        scraped_at=datetime.utcnow(),
        status="success" if all_products else "error",
        error=None if all_products else "No products extracted",
    )
